bugbug/models/duplicate.py

```diff
@@ -0,0 +1,129 @@
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from sklearn.compose import ColumnTransformer
+from sklearn.feature_extraction import DictVectorizer
+from sklearn.pipeline import Pipeline
+import shap
+from bugbug import bug_features, bugzilla, feature_cleanup
+import pandas as pd
+from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer, TfidfVectorizer
+import scipy
+from sklearn.metrics import accuracy_score
+import numpy as np
+from sklearn.model_selection import train_test_split, cross_validate
+from sklearn.svm import LinearSVC
+
+
+
+class DuplicateModel:
+    def __init__(self, lemmatization=False):
+        self.cross_validation_enabled = False
+        self.sampler = None
+
+        self.calculate_importance = True
+        self.pipeline = Pipeline([
+            ("tfidf", TfidfVectorizer(preprocessor=self.transform)),
+            ("clf", LinearSVC())
+        ])
+
+    def transform(self, X):
+        cleanup_functions = [
+            feature_cleanup.responses(),
+            feature_cleanup.hex(),
+            feature_cleanup.dll(),
+            feature_cleanup.fileref(),
+            feature_cleanup.url(),
+            feature_cleanup.synonyms(),
+            feature_cleanup.crash(),
+        ]
+        for cleanup_function in cleanup_functions:
+            X = cleanup_function(X)
+        return X
+
+    def get_text(self, bug):
+        return bug["summary"] + " ".join(comment["text"] for comment in bug["comments"])
+    
+    def get_duplicate_map(self):
+        
+        bugs = bugzilla.get_bugs()
+        bug = {}
+        duplicate_bug_map = {}
+        count = 500
+        for bug_data in bugs:
+            bug_id = int(bug_data["id"])
+
+            if bug_data["resolution"] == "DUPLICATE":
+                count -= 1
+                duplicate_bug_map[bug_id] = int(bug_data["dupe_of"])
+            
+            bug[bug_id] = bug_data
+            if not count :
+                break
+
+        return bug , duplicate_bug_map
+
+    def get_labels(self):
+        
+        bugs, duplicate_map = self.get_duplicate_map()
+
+        res = pd.DataFrame()
+
+        text1 = []
+        text2 = []
+
+        for ind, val in duplicate_map.items():
+            if val in bugs:
+                text1.append(self.get_text(bugs[ind]))
+                text2.append(self.get_text(bugs[val]))
+
+        l = len(text1)
+        count = 500
+        for ind, val in bugs.items():
+            if not count:
+                break
+            if "duplicate" not in val["resolution"]:
```

Comment:
The bug to which all bugs are duplicated is still part of the duplicate group, even if it is not resolved as "duplicate".
Here it would be better to just skip bugs which are in the duplicate_map

Reply:
Note: say we have bug 1, bug 2, bug 3, bug 4, bug 5.
bug 1 and bug 2 are duplicate.
bug 3 and bug 4 are duplicate.
bug 5 is not a duplicate of anything else.

When training, we need the following labels:
1 & 2 -> 1
1 & 3 -> 0
1 & 4 -> 0
1 & 5 -> 0
2 & 3 -> 0
2 & 4 -> 0
2 & 5 -> 0
3 & 4 -> 1
3 & 5 -> 0
4 & 5 -> 0